### PR TITLE
mfd: Add a driver for the RPI RP2040 GPIO bridge

### PR DESCRIPTION
**Pull request by roliver-rpi**
The Raspberry Pi RP2040 GPIO bridge is an I2C-attached MFD exposing both a Tx-only SPI controller, and a GPIO controller.

Due to the relative difference in transfer rates between standard-mode I2C and SPI, the GPIO bridge makes use of 12 MiB of non-volatile storage to cache repeated transfers. This cache is arranged in ~8 KiB blocks and is addressed by the MD5 digest of the data contained therein.

Optionally, this driver is able to take advantage of Raspberry Pi RP1 GPIOs to achieve faster than I2C data transfer rates.

### CODE REVIEW

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 1]
```diff
@@ -0,0 +1,1155 @@
...
+	priv_data->shash_desc->tfm = priv_data->shash;
+
+	controller->bus_num = np ? -1 : client->adapter->nr;
+	controller->num_chipselect = 1;
+	controller->mode_bits = SPI_CPOL | SPI_CPHA;
+	controller->bits_per_word_mask = SPI_BPW_MASK(8);
+	controller->min_speed_hz = 35000000;
+	controller->max_speed_hz = 35000000;
+	controller->max_transfer_size = rp2040_gbdg_max_transfer_size;
+	controller->max_message_size = rp2040_gbdg_max_transfer_size;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Seeing as `rp2040_gbdg_max_transfer_size` just returns `MAX_TRANSFER_SIZE`, do we need a function for it?

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 2]
```diff
@@ -0,0 +1,1155 @@
...
+
+	if (!priv_data->shash_desc) {
+		crypto_free_shash(priv_data->shash);
+		return dev_err_probe(
+			dev, ENOMEM,
+			"Could not allocate memory for shash_desc\n");
+	}
+	priv_data->shash_desc->tfm = priv_data->shash;
+
+	controller->bus_num = np ? -1 : client->adapter->nr;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Why is this conditional on `np` (aka `dev->of_node`)? It's only accessing `client->adapter`

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 3]
```diff
@@ -0,0 +1,1155 @@
...
+		priv_data->fast_xfer_gpios = NULL;
+		goto cleanup;
+	}
+
+cleanup:
+	if (of_args[0].np)
+		of_node_put(of_args[0].np);
+	if (of_args[1].np)
+		of_node_put(of_args[1].np);
+	return 0;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: So this is a success path? `cleanup` implied to me at least that it was an error path. `node_put` or similar would be more obvious.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 4]
```diff
@@ -0,0 +1,1155 @@
...
+
+	i2c_set_clientdata(client, priv_data);
+
+	priv_data->shash = crypto_alloc_shash("md5", 0, 0);
+	if (IS_ERR(priv_data->shash))
+		return dev_err_probe(dev, PTR_ERR(priv_data->shash),
+				     "Could not allocate shash\n");
+
+	if (crypto_shash_digestsize(priv_data->shash) != MD5_DIGEST_SIZE) {
+		crypto_free_shash(priv_data->shash);
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Repeated multiple times. It might be nice to make it a common exit path at the end of the function.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 5]
```diff
@@ -0,0 +1,1155 @@
...
+	ret = regulator_disable(reg);
+	if (ret)
+		return dev_err_probe(dev, ret, "could not disable regulator\n");
+
+	devm_pm_runtime_enable(dev);
+	pm_runtime_use_autosuspend(dev);
+
+	controller = devm_spi_alloc_master(dev, sizeof(struct rp2040_gbdg));
+	if (!controller)
+		return dev_err_probe(dev, ENOMEM,
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: This (and subsequent) error path(s) don't clean up pm_runtime.

If `devm_pm_runtime_enable` does pass enable across to devm, then disabling in _remove feels unbalanced.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 6]
```diff
@@ -0,0 +1,1155 @@
...
+static void rp2040_gbdg_gpio_set(struct gpio_chip *gc, unsigned int offset,
+				 int value)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	u32 pattern;
+	int ret;
+
+	if (offset >= NUM_GPIO) {
+		ret = -EINVAL;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: You do nothing with `ret`.
```
if (offset >= NUM_GPIO)
   return;
```
and no need for `else`.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 7]
```diff
@@ -0,0 +1,1155 @@
...
+}
+
+static int rp2040_gbdg_gpio_dir_in(struct gpio_chip *gc, unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	int ret;
+
+	if (offset >= NUM_GPIO) {
+		ret = -EINVAL;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]:
```
   return -EINVAL;
```
and no else.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 8]
```diff
@@ -0,0 +1,1155 @@
...
+}
+
+static int rp2040_gbdg_gpio_get_direction(struct gpio_chip *gc,
+					  unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	int ret;
+
+	if (offset >= NUM_GPIO)
+		ret = -EINVAL;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: `return -EINVAL`

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 9]
```diff
@@ -0,0 +1,1155 @@
...
+	}
+	dev_info(dev, "%s() found dev ID: %llx, fw ver. %u\n", __func__,
+		 info.id, info.version);
+
+	ret = regulator_disable(reg);
+	if (ret)
+		return dev_err_probe(dev, ret, "could not disable regulator\n");
+
+	devm_pm_runtime_enable(dev);
+	pm_runtime_use_autosuspend(dev);
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Do you want just autosuspend, or `pm_runtime_set_autosuspend_delay` to defer powerdown slightly in case the device is about to be used again? (See drivers/media/i2c/imx290.c for a full user).

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 10]
```diff
@@ -0,0 +1,1155 @@
...
+	int ret;
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	pattern = (1 << (offset + 8));
+	if (pattern & priv_data->gpio_requested)
+		return -EBUSY;
+
+	// Resume if previously no gpio requested
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Normal `/*   */` instead of `//`
Other places too, but this was the first I noticed.

`checkpatch.pl --strict` isn't a bad policy to adopt.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 11]
```diff
@@ -0,0 +1,1155 @@
...
+{
+	int ret;
+
+	if (priv_data->fast_xfer_gpios) {
+		ret = rp2040_gbdg_10byte_cmd(
+			priv_data->client, CMD_DAT_EMIT,
+			priv_data->fast_xfer_recv_gpio_base, length);
+		return ret ? ret :
+			     rp2040_gbdg_fast_xfer(priv_data, data, length);
+	} else {
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: You've returned in the first clause, so no point in an `else`

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 12]
```diff
@@ -0,0 +1,1155 @@
...
+#define HALF_BUFFER (4U * ONE_KIB)
+
+#define STATUS_SIZE 4
+#define MD5_DIGEST_SIZE 16
+#define VERSION_SIZE 4
+#define ID_SIZE 8
+#define TOTAL_RD_HDR_SIZE \
+	(STATUS_SIZE + MD5_DIGEST_SIZE + VERSION_SIZE + ID_SIZE)
+
+typedef u8 md5_digest[MD5_DIGEST_SIZE];
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: typedefs are frowned upon in the kernel coding standards.

I don't immediately see it being used other than in `struct rp2040_gbdg_device_info` either.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 13]
```diff
@@ -0,0 +1,1155 @@
...
+#define WRITE_DATA_PREFIX_SIZE 1
+
+#define FIXED_SIZE_CMD_PREFIX 0x81
+
+#define WRITE_DATA_UPPER_PREFIX 0x82
+#define WRITE_DATA_UPPER_PREFIX_SIZE 1
+
+#define NUM_GPIO 24
+
+enum FIXED_SIZE_COMMANDS {
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Naming of enums shouldn't have capitals, and should generally identify the module.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 14]
```diff
@@ -0,0 +1,1155 @@
...
+				       struct rp2040_gbdg_device_info *info)
+{
+	int ret;
+	u8 buf[TOTAL_RD_HDR_SIZE];
+	u8 *read_pos = buf;
+	size_t field_size;
+
+	do {
+		ret = i2c_master_recv(client, buf, sizeof(buf));
+	} while (ret == -ETIMEDOUT);
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: If I tie SCL low, do we ever get out of this loop?

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 15]
```diff
@@ -0,0 +1,1155 @@
...
+	field_size = sizeof_field(struct rp2040_gbdg_device_info, id);
+	memcpy(&info->id, read_pos, field_size);
+
+	return 0;
+}
+
+static int rp2040_gbdg_poll_device_info(struct i2c_client *client,
+					struct rp2040_gbdg_device_info *info)
+{
+	int ret;
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Many kernel subsystems will request "inverted Christmas tree", ie longest lines first when defining variables.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 16]
```diff
@@ -0,0 +1,1155 @@
...
+{
+	struct i2c_client *client = priv_data->client;
+	int ret;
+
+	if (len > RP2040_GBDG_BLOCK_SIZE)
+		return -EMSGSIZE;
+
+	u8 local_digest[MD5_DIGEST_SIZE];
+	u8 remote_digest[MD5_DIGEST_SIZE];
+	u8 ascii_digest[MD5_DIGEST_SIZE * 2 + 1] = { 0 };
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Variable declarations before all code.

File: drivers/mfd/rpi-rp2040-gpio-bridge.c [hunk 17]
```diff
@@ -0,0 +1,1155 @@
...
+
+	gpiod_direction_output(priv_data->fast_xfer_gpios->desc[0], 1);
+	gpiod_direction_output(priv_data->fast_xfer_gpios->desc[1], 0);
+
+	rp2040_gbdg_rp1_calc_offsets(priv_data->fast_xfer_data_index,
+				     &data_bank, &data_offset);
+	rp2040_gbdg_rp1_calc_offsets(priv_data->fast_xfer_clock_index,
+				     &clock_bank, &clock_offset);
+
+	void __iomem *data_set =
```
Comment (6by9) [inline drivers/mfd/rpi-rp2040-gpio-bridge.c]: Variable declaration in the middle of a function.

File: arch/arm/boot/dts/overlays/rpi-rp2040-gpio-bridge.dtsi [hunk 18]
```diff
@@ -0,0 +1,20 @@
+spi_bridge: spi@40 {
```
Comment (6by9) [inline arch/arm/boot/dts/overlays/rpi-rp2040-gpio-bridge.dtsi]: dtoverlay should be a separate patch to driver or bindings.

File: Documentation/devicetree/bindings/mfd/raspberrypi,rp2040-gpio-bridge.yaml [hunk 19]
```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: GPL-2.0-only
```
Comment (6by9) [inline Documentation/devicetree/bindings/mfd/raspberrypi,rp2040-gpio-bridge.yaml]: Bindings should be a separate patch to anything else.

Does it validate? `make dt_binding_check`

File: drivers/mfd/Kconfig [hunk 20]
```diff
@@ -2362,5 +2362,14 @@ config MFD_RSMU_SPI
 	  Additional drivers must be enabled in order to use the functionality
 	  of the device.
 
+config MFD_RASPBERRYPI_RP2040_GPIO_BRIDGE
```
Comment (6by9) [inline drivers/mfd/Kconfig]: Is it an MFD in the kernel sense? Most MFD drivers allow for a parent that can acquire shared resources, and then you have separate drivers off it for the specific functions.

bcm2835-pm parents watchdog and power drivers
POE+ HAT uses simple-mfd-i2c as the core, with a PWM and power supply driver on top.

**Review by 6by9 — COMMENTED** [review]
Largely coding style niggles.

General minor clean up of whether there is a space before the final return of a function would be nice (it's generally expected).

**Comment by roliver-rpi** [comment]
Any further comments on this?

I believe all code review comments have been addressed.

Perhaps the single remaining outstanding question is whether or not we're happy with this in mfd?

Thoughts on potential alternate locations @pelwell @6by9 @naushir ?

If we're happy with mfd then I believe this is ready to merge.

**Comment by pelwell** [comment]
> whether or not we're happy with this in mfd?

What are the alternatives:
* An I2C attached SPI controller would live in `drivers/spi`. 
* Many devices also provide GPIOs. For example, sc16is75x is an I2C- or SPI-attached UART that also acts as a GPIO controller. It lives in `drivers/tty/serial`, again suggesting that `drivers/spi` is appropriate.
* Shove it in `drivers/misc`?

**Comment by roliver-rpi** [comment]
Driver has been moved to drivers/spi

File: drivers/spi/spi-rp2040-gpio-bridge.c [hunk 21]
```diff
@@ -0,0 +1,1219 @@
...
+
+	if (IS_ERR(reg))
+		return PTR_ERR(reg);
+
+	rp2040_gbdg->regulator = reg;
+
+	return 0;
+}
+
+static void rp2040_gbdg_parse_dt(struct rp2040_gbdg *rp2040_gbdg)
```
Comment (pelwell) [inline drivers/spi/spi-rp2040-gpio-bridge.c:908]: Is it reasonable that this has no return value? It has many points where it returns early after an error.

**Review by pelwell — COMMENTED** [review]
Realistically I'm not going to fully grok this new driver, but apart from the lack of error return from the DT parsing it looks OK to me.

File: drivers/spi/spi-rp2040-gpio-bridge.c (hunk 21 above)
Comment (roliver-rpi) [reply drivers/spi/spi-rp2040-gpio-bridge.c:908]: Only the parsing of the 'bypass-cache' property in rp2040_gbdg_parse_dt() is required for all platforms. This cannot fail due to the API of of_property_read_bool() returning the bool.

Any failure in the rest of rp2040_gbdg_parse_dt() is not an issue as the driver will use I2C to transfer blocks of data as a fallback (rather than fast-xfer using RP1 bit-bashing). In fact, we expect failure on platforms without RP1.

If we were to return a value from rp2040_gbdg_parse_dt, in it's current form, it would only be informational and not to be acted upon (other than perhaps a dev_info).

I'm happy to refactor rp2040_gbdg_parse_dt() if we can think of a cleaner/easier to understand API. Alternatively, I could comment the function as it's perhaps not sufficiently clear in its current form.

Comment (pelwell) [reply drivers/spi/spi-rp2040-gpio-bridge.c:908]: It's OK - if you're happy justifying it being as it is now, I'm happy to merge it
